File: Design-Patterns-Using-CSharp/Factory/Factory1.cs
using System;

// Product interface
interface IProduct
{
    void Operation();
}

// Concrete Product classes
class ConcreteProduct1 : IProduct
{
    public void Operation()
    {
        Console.WriteLine("Concrete Product1 Operation!");
    }
}

class ConcreteProduct2 : IProduct
{
    public void Operation()
    {
        Console.WriteLine("Concrete Product2 Operation!");
    }
}

// Creator abstract class
abstract class Factory
{
    public abstract IProduct CreateProduct();
}

// Concrete Creator classes
class ConcreteFactory1 : Factory
{
    public override IProduct CreateProduct()
    {
        return new ConcreteProduct1();
    }
}

class ConcreteFactory2 : Factory
{
    public override IProduct CreateProduct()
    {
        return new ConcreteProduct2();
    }
}

// Client code
class Factory1
{
    static void Main(string[] args)
    {
        Factory factory1 = new ConcreteFactory1();
        IProduct product1 = factory1.CreateProduct();
        product1.Operation();

        Factory factory2 = new ConcreteFactory2();
        IProduct product2 = factory2.CreateProduct();
        product2.Operation();
    }
}

/*
Concrete Product1 Operation!
Concrete Product2 Operation!
*/

File: Design-Patterns-Using-Java/Factory/Factory1.java
// Product interface
interface Product {
    void operation();
}

// Concrete Product classes
class ConcreteProduct1 implements Product {
    @Override
    public void operation() {
        System.out.println("Concrete Product1 Operation!");
    }
}

class ConcreteProduct2 implements Product {
    @Override
    public void operation() {
        System.out.println("Concrete Product2 Operation!");
    }
}

// Creator abstract class
abstract class Factory {
    public abstract Product createProduct();
}

// Concrete Creator classes
class ConcreteFactory1 extends Factory {
    @Override
    public Product createProduct() {
        return new ConcreteProduct1();
    }
}

class ConcreteFactory2 extends Factory {
    @Override
    public Product createProduct() {
        return new ConcreteProduct2();
    }
}

// Client code
public class Factory1 {
    public static void main(String[] args) {
        Factory factory1 = new ConcreteFactory1();
        Product product1 = factory1.createProduct();
        product1.operation();

        Factory factory2 = new ConcreteFactory2();
        Product product2 = factory2.createProduct();
        product2.operation();
    }
}


/* 
Concrete Product1 Operation!
Concrete Product2 Operation!
*/
